File: javascript/src/visitor/visitor.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ConcreteElementA } from "./concrete-element-a.js";
import { ConcreteElementB } from "./concrete-element-b.js";
import { ConcreteVisitor1 } from "./concrete-visitor1.js";
import { ConcreteVisitor2 } from "./concrete-visitor2.js";
import { ObjectStructure } from "./object-structure.js";

function structure() {
  const s = new ObjectStructure();
  for (const element of [new ConcreteElementA(), new ConcreteElementB(), new ConcreteElementA()]) {
    s.add(element);
  }
  return s;
}

describe("Visitor", () => {
  it("each element dispatches to the visit method for its class", () => {
    const visitor = new ConcreteVisitor1();
    new ConcreteElementB().accept(visitor);
    assert.equal(visitor.result(), "visited ConcreteElementB");
    const other = new ConcreteVisitor2();
    new ConcreteElementA().accept(other);
    assert.equal(other.result(), "A");
  });

  it("visitors accumulate over the whole structure", () => {
    const visitor = new ConcreteVisitor2();
    structure().accept(visitor);
    assert.equal(visitor.result(), "A+B+A");
  });

  it("an empty structure yields an empty result", () => {
    const visitor = new ConcreteVisitor1();
    new ObjectStructure().accept(visitor);
    assert.equal(visitor.result(), "");
  });

  it("a new operation needs no change to the elements", () => {
    const counts = [];
    const counter = { visitConcreteElementA: () => counts.push("a"), visitConcreteElementB: () => counts.push("b") };
    structure().accept(counter);
    assert.deepEqual(counts, ["a", "b", "a"]);
  });
});

File: java/src/test/java/com/penapereira/patterns/visitor/VisitorTest.java
package com.penapereira.patterns.visitor;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class VisitorTest {

    private static ObjectStructure structure() {
        ObjectStructure structure = new ObjectStructure();
        structure.add(new ConcreteElementA());
        structure.add(new ConcreteElementB());
        structure.add(new ConcreteElementA());
        return structure;
    }

    @Test
    void eachElementDispatchesToTheVisitMethodForItsClass() {
        ConcreteVisitor1 visitor = new ConcreteVisitor1();
        new ConcreteElementB().accept(visitor);
        assertEquals("visited ConcreteElementB", visitor.result());
        ConcreteVisitor2 other = new ConcreteVisitor2();
        new ConcreteElementA().accept(other);
        assertEquals("A", other.result());
    }

    @Test
    void visitorsAccumulateOverTheWholeStructure() {
        ConcreteVisitor2 visitor = new ConcreteVisitor2();
        structure().accept(visitor);
        assertEquals("A+B+A", visitor.result());
    }

    @Test
    void anEmptyStructureYieldsAnEmptyResult() {
        ConcreteVisitor1 visitor = new ConcreteVisitor1();
        new ObjectStructure().accept(visitor);
        assertEquals("", visitor.result());
    }

    @Test
    void aNewOperationNeedsNoChangeToTheElements() {
        List<String> counts = new ArrayList<>();
        Visitor counter = new Visitor() {
            @Override
            public void visitConcreteElementA(ConcreteElementA element) {
                counts.add("a");
            }

            @Override
            public void visitConcreteElementB(ConcreteElementB element) {
                counts.add("b");
            }
        };
        structure().accept(counter);
        assertEquals(List.of("a", "b", "a"), counts);
    }
}
